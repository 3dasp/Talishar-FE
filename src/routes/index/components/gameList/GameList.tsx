import React, { useState } from 'react';
import { useGetGameListQuery } from 'features/api/apiSlice';
import styles from './GameList.module.css';
import InProgressGame from '../inProgressGame';
import OpenGame from '../openGame';
import Filter from '../filter';
import { GAME_FORMAT } from 'constants';
import FormatList from '../formatList';
import { useAutoAnimate } from '@formkit/auto-animate/react';

export interface IOpenGame {
  p1Hero?: string;
  format: string;
  formatName?: string;
  description?: string;
  gameName: number;
}

export interface IGameInProgress {
  p1Hero?: string;
  p2Hero?: string;
  gameName: number;
  secondsSinceLastUpdate?: number;
}

export interface GameListResponse {
  gamesInProgress: IGameInProgress[];
  openGames: IOpenGame[];
  canSeeQueue?: boolean;
}

const GAME_LIST_POLLING_INTERVAL = 10000; // in ms

const GameList = () => {
  const { data, isLoading, error, refetch, isFetching } = useGetGameListQuery(
    {},
    {}
  );

  const [heroFilter, setHeroFilter] = useState<string[]>([]);
  const [formatFilter, setFormatFilter] = useState<string | null>(null);

  const [parent] = useAutoAnimate();

  let sortedOpenGames = data?.openGames ? [...data.openGames] : [];
  sortedOpenGames = sortedOpenGames
    .filter((game: IOpenGame) => {
      return (
        heroFilter.length === 0 ||
        heroFilter.find((hero) => hero === game.p1Hero)
      );
    })
    .filter((game: IOpenGame) => {
      return formatFilter === null || game.format === formatFilter;
    })
    .sort((a, b) => a.format.localeCompare(b.format));

  let filteredGamesInProgress = data?.gamesInProgress
    ? [...data.gamesInProgress]
    : [];
  filteredGamesInProgress = filteredGamesInProgress
    .filter((game) => {
      return (
        heroFilter.length === 0 ||
        heroFilter.find((hero) => {
          return hero === game.p1Hero || hero === game.p2Hero;
        })
      );
    })
    .sort((a, b) => {
      return (a.secondsSinceLastUpdate ?? 0) - (b.secondsSinceLastUpdate ?? 0);
    });

  const handleReloadClick = () => {
    refetch();
  };

  const otherFormats = [
    GAME_FORMAT.OPEN_FORMAT,
    GAME_FORMAT.COMMONER,
    GAME_FORMAT.CLASH
  ];
  const blitz = sortedOpenGames.filter(
    (game) => game.format === GAME_FORMAT.BLITZ
  );
  const compBlitz = sortedOpenGames.filter(
    (game) => game.format === GAME_FORMAT.COMPETITIVE_BLITZ
  );
  const cc = sortedOpenGames.filter(
    (game) => game.format === GAME_FORMAT.CLASSIC_CONSTRUCTED
  );
  const compcc = sortedOpenGames.filter(
    (game) => game.format === GAME_FORMAT.COMPETITIVE_CC
  );
  const otherGames = sortedOpenGames.filter((game) =>
    otherFormats.includes(game.format)
  );

  return (
    <article className={styles.gameList}>
      <div className={styles.titleDiv}>
        <h1 className={styles.title}>Games</h1>
        <button
          onClick={handleReloadClick}
          aria-busy={isFetching}
          disabled={isFetching}
          className={styles.reloadButton}
        >
          Reload
        </button>
      </div>
      {isLoading ? <div aria-busy="true">Loading!</div> : null}
      {error ? <div>ERROR!</div> : null}
      {!isLoading && !error && <Filter setHeroFilter={setHeroFilter} />}
      <FormatList gameList={blitz} name="Blitz" />
      <FormatList gameList={compBlitz} name="Competitive Blitz" />
      <FormatList gameList={cc} name="Classic Constructed" />
      <FormatList gameList={compcc} name="Comp CC" />
      <FormatList gameList={otherGames} name="Other" isOther />
      {data != undefined && (
        <div data-testid="games-in-progress" ref={parent}>
          <h5 className={styles.subSectionTitle}>
            Games in progress: <span>({data.gamesInProgress.length})</span>
          </h5>
          {filteredGamesInProgress.map((entry, ix: number) => {
            return (
              <InProgressGame entry={entry} ix={ix} key={entry.gameName} />
            );
          })}
        </div>
      )}
    </article>
  );
};

export default GameList;
