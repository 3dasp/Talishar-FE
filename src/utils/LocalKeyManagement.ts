export const saveGameAuthKey = (gameId: number, authKey: string): void => {
  if (gameId <= 0) {
    console.error('Invalid game ID, cannot save authKey for game:', gameId);
    return;
  }
  sessionStorage.setItem(String(gameId), authKey);
};

export const loadGameAuthKey = (gameId: number): string => {
  if (gameId <= 0) {
    console.error('Invalid game ID, cannot load authKey for game:', gameId);
    return '';
  }

  const localValue = sessionStorage.getItem(String(gameId));
  if (!localValue) {
    console.error(
      'Attempted to access an authKey of a game that does not exist, game:' +
        gameId
    );
  }
  return localValue || '';
};

export const deleteGameAuthKey = (gameId: number): void => {
  if (gameId <= 0) {
    console.error('Invalid game ID, cannot delete authKey for game:', gameId);
    return;
  }

  const valueExists = sessionStorage.getItem(String(gameId));
  if (!valueExists) {
    console.error(
      'Attempted to delete a authKey of a game that does not exist, game',
      gameId
    );
    return;
  }
  sessionStorage.removeItem(String(gameId));
};
