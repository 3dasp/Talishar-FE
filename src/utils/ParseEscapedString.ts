const CARDRE = /{{(.+?)\|(.+?)\|(.+?)}}/g;

export const replaceText = (inputString: string): string => {
  const mapping: { [key: string]: string } = {
    '0': 'gray',
    '1': 'red',
    '2': 'yellow',
    '3': 'blue'
  };

  return inputString.replace(CARDRE, (match, p1, p2, p3) => {
    const color = mapping[p3];
    const imgPath = `./WebpImages/${p1}.webp`;
    return `<span onmouseover="ShowDetail(event, '${imgPath}')" onmouseout="HideCardDetail()" style="color:${color}">${p2}</span>`;
  });
};
