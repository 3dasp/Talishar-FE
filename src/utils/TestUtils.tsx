/* eslint-disable @typescript-eslint/ban-types */
import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import { AppStore, RootState, setupStore } from '../app/Store';

import InitialGameState from '../features/game/InitialGameState';
import { MemoryRouter } from 'react-router-dom';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    // @ts-ignore - I know I know...
    preloadedState = {
      game: InitialGameState
    },
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <MemoryRouter>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
