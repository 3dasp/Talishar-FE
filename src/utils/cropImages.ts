import { CLOUD_IMAGES_URL } from "appConstants";

export const generateCroppedImageUrl = (cardName: string): string => `${CLOUD_IMAGES_URL}/crops/${cardName}_cropped.png`;