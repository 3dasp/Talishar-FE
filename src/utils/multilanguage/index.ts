export {
  getCollectionCardImagePath,
  getCardBackImagePath,
  loadInitialLanguage
} from './multilanguage';
export type { pathType } from './types';
export {
  DEFAULT_LANGUAGE,
  CARD_IMAGES_PATH,
  CARD_SQUARES_PATH
} from './constants';
