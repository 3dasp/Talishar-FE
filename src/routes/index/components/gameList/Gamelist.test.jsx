import { renderWithProviders } from 'utils/TestUtils';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

class ResizeObserver {
  observe() {
    // do nothing
  }
  unobserve() {
    // do nothing
  }
  disconnect() {
    // do nothing
  }
}

window.ResizeObserver = ResizeObserver;

import GameList from './GameList';

it('displays loading! when loading', () => {
  console.log('hello');
  const { container } = renderWithProviders(<GameList />);
  expect(container).toMatchSnapshot();
});

// TODO: Rewrite the test so it works properly
it('displays the list of games! when loaded', async () => {
  const div = renderWithProviders(<GameList />);
  const list = await div.findByTestId('games-in-progress');
  expect(list).toMatchSnapshot();
});
