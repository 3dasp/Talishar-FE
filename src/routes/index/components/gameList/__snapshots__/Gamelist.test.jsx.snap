// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`displays loading! when loading 1`] = `
<div>
  <article
    class="_gameList_4a28c4"
  >
    <div
      class="_titleDiv_4a28c4"
    >
      <h1
        class="_title_4a28c4"
      >
        Games
      </h1>
      <button
        aria-busy="true"
        class="_reloadButton_4a28c4"
        disabled=""
      >
        Reload
      </button>
    </div>
    <div
      aria-busy="true"
    >
      Loading!
    </div>
  </article>
</div>
`;

exports[`displays the list of games! when loaded 1`] = `
<div
  data-testid="games-in-progress"
>
  <h5
    class="_subSectionTitle_4a28c4"
  >
    Games in progress: 
    <span>
      (
      3
      )
    </span>
  </h5>
  <div
    class="_gameItem_0a3e89"
  >
    <div>
      <img
        src="/crops/ELE062_cropped.png"
      />
    </div>
     
    -
     
    <div>
      <img
        src="/crops/UPR044_cropped.png"
      />
    </div>
    <div>
      <button
        class="_button_0a3e89 secondary"
      >
        Spectate
      </button>
    </div>
  </div>
  <div
    class="_gameItem_0a3e89"
  >
    <div>
      <img
        src="/crops/WTR001_cropped.png"
      />
    </div>
     
    -
     
    <div>
      <img
        src="/crops/UPR001_cropped.png"
      />
    </div>
    <div>
      <button
        class="_button_0a3e89 secondary"
      >
        Spectate
      </button>
    </div>
  </div>
  <div
    class="_gameItem_0a3e89"
  >
    <div>
      <img
        src="/crops/UPR102_cropped.png"
      />
    </div>
     
    -
     
    <div>
      <img
        src="/crops/UPR044_cropped.png"
      />
    </div>
    <div>
      <button
        class="_button_0a3e89 secondary"
      >
        Spectate
      </button>
    </div>
  </div>
</div>
`;
